util/src/journaldb/overlayrecentdb.rs

```diff
@@ -430,8 +434,51 @@ impl HashDB for OverlayRecentDB {
 		v.or_else(|| self.payload(key))
 	}
 
+	fn get_exec<'this>(
+		&'this self,
+		key: &'this H256,
+		f: &'this mut for<'a: 'this> FnMut(&'a [u8]),
+	) {
+		use ::std::cell::Cell;
+
+		let k = self.transaction_overlay.raw_ref(key);
+
+		if let Some((d, rc)) = k {
+			if rc > 0 {
+				f(d.as_ref());
+				return;
+			}
+		}
+
+		let was_called = Cell::new(false);
+
+		// This type annotation is necessary, the inferred lifetime is too restrictive otherwise.
+		// Possible bug in the Rust compiler.
+		let mut wrapper = |d: &[u8]| {
+			was_called.set(true);
+			f(d);
```

Comment:
these are the assumptions I am operating on:
  1. `'this` (lifetime of self borrow) can be safely coerced to a smaller lifetime, but must cover the entire function body
  2. the read guard itself may live for `'this' (why you can return a `ReadGuard<'this, T>`)
  3. borrows of `ReadGuard<'this, T>` will be strictly smaller than `'this` (why you can't return `&'a [u8]` from a `ReadGuard<'this, Vec<u8>>`)

But somehow we are taking data which is borrowed from an internal `ReadGuard` and calling a function that says "I only take data that lives at least as long as 'this" with it.